cluster-autoscaler/cloudprovider/openstack/openstack_manager_heat.go

```diff
@@ -0,0 +1,299 @@
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog"
+	provider_os "k8s.io/kubernetes/pkg/cloudprovider/providers/openstack"
+)
+
+const (
+	stackStatusUpdateInProgress = "UPDATE_IN_PROGRESS"
+	stackStatusUpdateComplete   = "UPDATE_COMPLETE"
+)
+
+// StatusesPreventingUpdate is a set of statuses that would prevent
+// the cluster from successfully scaling.
+var StatusesPreventingUpdate = sets.NewString(
+	clusterStatusUpdateInProgress,
+	clusterStatusUpdateFailed,
+)
+
+// OpenstackManagerHeat implements the OpenstackManager interface.
+//
+// Most interactions with the cluster are done directly with magnum,
+// but scaling down requires an intermediate step using heat to
+// delete the specific nodes that the autoscaler has picked for removal.
+type OpenstackManagerHeat struct {
+	clusterClient *gophercloud.ServiceClient
+	heatClient    *gophercloud.ServiceClient
+	clusterName   string
+
+	stackName string
+	stackID   string
+
+	waitTimeStep time.Duration
+}
+
+// CreateOpenstackManagerHeat sets up cluster and stack clients and returns
+// an OpenstackManagerHeat.
+func CreateOpenstackManagerHeat(configReader io.Reader, discoverOpts cloudprovider.NodeGroupDiscoveryOptions, opts config.AutoscalingOptions) (*OpenstackManagerHeat, error) {
+	var cfg provider_os.Config
+	if configReader != nil {
+		if err := gcfg.ReadInto(&cfg, configReader); err != nil {
+			klog.Errorf("Couldn't read config: %v", err)
+			return nil, err
+		}
+	}
+
+	if opts.ClusterName == "" {
+		klog.Fatalf("The cluster-name parameter must be set")
+	}
+
+	authOpts := toAuthOptsExt(cfg)
+
+	provider, err := openstack.NewClient(cfg.Global.AuthURL)
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate client: %v", err)
+	}
+
+	if cfg.Global.CAFile != "" {
+		roots, err := certutil.NewPool(cfg.Global.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		config := &tls.Config{}
+		config.RootCAs = roots
+		provider.HTTPClient.Transport = netutil.SetOldTransportDefaults(&http.Transport{TLSClientConfig: config})
+
+	}
+
+	err = openstack.AuthenticateV3(provider, authOpts, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("could not authenticate: %v", err)
+	}
+
+	clusterClient, err := openstack.NewContainerInfraV1(provider, gophercloud.EndpointOpts{Type: "container-infra", Name: "magnum", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create container-infra client: %v", err)
+	}
+
+	heatClient, err := openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{Type: "orchestration", Name: "heat", Region: cfg.Global.Region})
+	if err != nil {
+		return nil, fmt.Errorf("could not create orchestration client: %v", err)
+	}
+
+	manager := OpenstackManagerHeat{
+		clusterClient: clusterClient,
+		clusterName:   opts.ClusterName,
+		heatClient:    heatClient,
+		waitTimeStep:  waitForStatusTimeStep,
+	}
+
+	manager.stackID, err = manager.GetStackID()
```

Comment:
why not just  manager.stackID, err:= clusters.Get(osm.clusterClient, osm.clusterName).Extract()

Reply:
I'll change it.

Reply:
Done in https://github.com/kubernetes/autoscaler/pull/1690/commits/0d7387f3806a8c9f48f60030214a57b2bf1282b6.